File: data/figures/chapter02/figure_2_4/plot_2_4.py
"""Reproduce Figure 2.4 from figure_2_4.csv only (NumPy and Matplotlib).

Static figure contract: compare four observed/reconstructed mesh-preparation
behaviours. Orange = beam centre-lines; grey = skin edges; blue = receiving
facets. Open/filled and dashed/solid distinguish states without colour alone.
Geometry uses orthonormal local bases, equal length scales and explicit units.
This is local postprocessing, not a solver run; no other data source is opened.
"""
import argparse
import csv
from pathlib import Path
import numpy as np
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d.art3d import Poly3DCollection

OUT=Path(__file__).resolve().parent
BLUE="#31639e"; ORANGE="#c7611f"; GREY="#929aA1"; INK="#2c3035"
plt.rcParams.update({"font.family":"DejaVu Sans","font.size":11,"text.color":INK,
                     "axes.titlecolor":INK,"figure.facecolor":"white"})


def subset(rows,**conditions):
    return [r for r in rows if all(r[k]==str(v) for k,v in conditions.items())]


def xyz(rows):
    return np.asarray([[float(r[c]) for c in ["x_m","y_m","z_m"]] for r in rows])


def one(rows,role,state):
    return subset(rows,entity="node",role=role,state=state)[0]


def point(rows,role,state):
    return xyz([one(rows,role,state)])[0]


def grouped(rows,entity):
    chosen=subset(rows,entity=entity)
    return [sorted(subset(chosen,element_id=e),key=lambda r:int(r["vertex_index"]))
            for e in sorted({r["element_id"] for r in chosen},key=int)]


def basis(rows,role="B",direction=None):
    r=one(rows,role,"original");p=point(rows,role,"projection")
    v=xyz(subset(rows,entity="skin_vertex",element_id=r["element_id"]))
    n=np.cross(v[1]-v[0],v[2]-v[0]);n/=np.linalg.norm(n)
    if np.dot(p-point(rows,role,"original"),n)<0:n=-n
    if direction is None:direction=point(rows,"C","original")-point(rows,"A","original")
    t=direction-np.dot(direction,n)*n;t/=np.linalg.norm(t)
    b=np.column_stack([t,np.cross(n,t),n])
    assert np.allclose(b.T@b,np.eye(3),atol=1e-12)
    return p,b


def label(ax,q,text,offset=(0.,-.3),color=INK):
    ax.text(q[0]+offset[0],q[1]+offset[1],text,color=color,fontsize=12,weight="bold")


def upper(data):
    fig=plt.figure(figsize=(14,6.8))
    axes=np.array([[fig.add_axes([.055+.475*c,.56-.45*r,.4,.30]) for c in range(2)] for r in range(2)])
    names=["single_B_A_ordinary_C_beyond","B_rejected_length"]
    titles=["(a) Single-node repair: original","(b) Single-node repair: final",
            "(c) Length guard: proposed correction","(d) Length guard: retained mesh"]
    for row,name in enumerate(names):
        d=subset(data,example=name);p,b=basis(d)
        transform=lambda a:(a-p)@b[:,[0,2]]*1000
        old=transform(np.array([point(d,r,"original") for r in "ABC"]))
        bounds=[old[:,0].min()-.75,old[:,0].max()+.75,old[:,1].min()-.7,.85]
        for col in range(2):
            ax=axes[row,col]
            for face in grouped(d,"skin_vertex"):
                q=transform(xyz(face));q=np.vstack([q,q[0]])
                receiving=any(r and "proposed_not_receiving" not in r for r in face[0]["receiving_for"].split(";"))
                ax.plot(*q.T,color=BLUE if receiving else GREY,lw=1.4 if receiving else .55,zorder=1)
            state="original" if row==0 and col==0 else "final"
            proposal=row==1 and col==0
            if proposal:state="all_proposals"
            q=transform(np.array([point(d,r,state) for r in "ABC"]))
            ax.plot(*q.T,"--" if proposal else "-",color=ORANGE,lw=3,zorder=3)
            ax.scatter(q[[0,2],0],q[[0,2],1],s=50,c=INK,edgecolor="white",zorder=4)
            if proposal:ax.scatter(q[1,0],q[1,1],s=70,marker="x",color=ORANGE,linewidth=1.5,zorder=5)
            else:ax.scatter(q[1,0],q[1,1],s=70,marker="o",facecolor=ORANGE if row==0 and col==1 else "white",edgecolor=ORANGE,linewidth=1.5,zorder=5)
            label(ax,q[0],"A",(-.3,-.22));label(ax,q[2],"C",(.1,-.1))
            label(ax,q[1],"B*" if proposal else ("B'" if row==0 and col==1 else "B"),(.12,-.4) if proposal else (-.42,-.12),ORANGE)
            ax.set_aspect("equal");ax.set_xlim(bounds[:2]);ax.set_ylim(bounds[2:]);ax.axis("off")
            ax.set_title(titles[row*2+col],fontsize=13,pad=15)
            x,y=bounds[0]+.15,bounds[2]+.2
            ax.plot([x,x+1],[y,y],color=INK,lw=1)
            ax.text(x+.5,y-.25,"1 mm",ha="center",fontsize=9)
    d=subset(data,example=names[0]);br=one(d,"B","original");cr=one(d,"C","original")
    note=f"B moves {float(br['coordinate_correction_m'])*1e3:.3f} mm; A remains coupled; C is outside repair reach ({float(cr['projection_distance_m'])*1e3:.3f} > {float(cr['snap_tolerance_m'])*1e3:.3f} mm)."
    fig.text(.055,.475,note,fontsize=11)
    d=subset(data,example=names[1]);ids={one(d,r,"original")["grid_id"] for r in "AB"}
    members=grouped(subset(d,state="all_proposals"),"beam_endpoint")
    ab=next(e for e in members if {r["grid_id"] for r in e}==ids)
    guard=float(next(r["value"] for r in data if r["metric"]=="snap_max_beam_length_change_fraction"))
    fig.text(.06,.025,f"B* is rejected: AB would lengthen by {float(ab[0]['fractional_length_change'])*100:.1f}%, exceeding the {guard*100:.0f}% guard. B remains unchanged.",fontsize=11)
    return fig


def lower(data):
    fig=plt.figure(figsize=(14,7.3))
    d=subset(data,example="two_node");p,b=basis(d)
    trans=lambda a:(a-p)@b*1000
    allq=trans(xyz([r for r in d if r["x_m"]]));limits=[(allq[:,0].min()-.3,allq[:,0].max()+.3),
        (allq[:,1].min()-.2,allq[:,1].max()+.2),(-3.15,.5)]
    for col,state in enumerate(["original","final"]):
        ax=fig.add_axes([.04+col*.5,.52,.42,.40],projection="3d",computed_zorder=False)
        for face in grouped(d,"skin_vertex"):
            receiving=bool(face[0]["receiving_for"]);q=trans(xyz(face))
            poly=Poly3DCollection([q],facecolor=BLUE if receiving else "#dce0e4",
                edgecolor=BLUE if receiving else GREY,alpha=.22 if receiving else .1,
                linewidth=1.2 if receiving else .7,zorder=1)
            ax.add_collection3d(poly)
            if receiving:
                centre=q.mean(axis=0);role=face[0]["receiving_for"]
                centre[2]+=.30 if role=="B" else -.60
                ax.text(*centre,"$T_"+role+"$",color=BLUE,fontsize=11,ha="center",zorder=5)
        for beam in grouped(subset(d,state=state),"beam_endpoint"):
            q=trans(xyz(beam));ax.plot(*q.T,color=ORANGE,lw=3.2,zorder=3)
        for role in "ABC":
            q=trans(point(d,role,state));marker="d" if role=="C" else "o"
            fill=INK if role=="A" else (ORANGE if state=="final" else "white")
            ax.scatter(*q,c=fill,edgecolor=ORANGE,s=45,marker=marker,depthshade=False,zorder=4)
            dz=-.4 if role=="B" and state=="final" else .2
            if role=="C" and state=="original":dz=-.2
            ax.text(q[0]+.08,q[1],q[2]+dz,role+("'" if state=="final" and role!="A" else ""),fontsize=12,weight="bold",zorder=5)
        if state=="original":
            q=trans(np.array([point(d,r,"projection") for r in "BC"]))
            ax.scatter(*q.T,c=BLUE,marker="x",s=30,depthshade=False,zorder=4)
        ax.set_xlim(limits[0]);ax.set_ylim(limits[1]);ax.set_zlim(limits[2])
        ax.set_box_aspect([hi-lo for lo,hi in limits],zoom=1.65);ax.set_proj_type("ortho")
        ax.view_init(elev=25,azim=-80);ax.set_axis_off()
        ax.plot([-2.4,-1.4],[0,0],[-2.7,-2.7],color=INK,lw=1.2)
        ax.text(-1.9,0,-2.97,"1 mm",fontsize=9,ha="center")
        ax.set_title("(e) Two adjacent nodes: before repair" if col==0 else "(f) Both nodes moved to their receiving facets",fontsize=13,pad=0)
    corrections=[float(one(d,r,"original")["coordinate_correction_m"])*1000 for r in "BC"]
    fig.text(.055,.475,f"B correction: {corrections[0]:.3f} mm; C correction: {corrections[1]:.3f} mm. Two receiving triangles are highlighted.",fontsize=11)
    d=subset(data,example="finite_offset");r=one(d,"Q","original");gid=r["grid_id"]
    ends=subset(d,entity="beam_endpoint",state="original")
    other=xyz([e for e in ends if e["grid_id"]!=gid]);p,b=basis(d,"Q",other[1]-other[0])
    transform=lambda a:(a-p)@b[:,[0,2]]*1e6
    gap=float(r["projection_distance_m"])*1e6
    for col,state in enumerate(["original","final"]):
        ax=fig.add_axes([.07+.5*col,.105,.365,.29])
        for face in grouped(d,"skin_vertex"):
            q=transform(xyz(face));q=np.vstack([q,q[0]]);ax.plot(*q.T,color=BLUE,lw=1.5)
        for beam in grouped(subset(d,state=state),"beam_endpoint"):
            ax.plot(*transform(xyz(beam)).T,color=ORANGE,lw=3.2)
        q=transform(point(d,"Q",state));ax.scatter(*q,facecolor="white",edgecolor=ORANGE,s=60,zorder=5)
        ax.scatter(0,0,c=BLUE,marker="x",s=45,zorder=5)
        ax.text(-3.2,-9,"Q",fontsize=13,weight="bold");ax.text(1,1.6,"Projection p",color=BLUE,fontsize=10)
        ax.text(-21,3.7,"Receiving skin element, edge-on",color=BLUE,fontsize=10)
        ax.text(4,-5.8,f"Gap: {gap:.3f} μm",fontsize=10)
        ax.plot([-21,-16],[-13,-13],color=INK,lw=1.3);ax.text(-18.5,-14.6,"5 μm",ha="center",fontsize=9)
        ax.set_aspect("equal");ax.set_xlim(-23,23);ax.set_ylim(-15,7);ax.axis("off")
        ax.set_title("(g) Near-skin node before coupling" if col==0 else "(h) Ordinary MPC: coordinates unchanged",fontsize=13,pad=15)
    fig.text(.055,.025,f"Uniform side-view crop: the {gap:.3f} μm gap remains below the 10 μm coupling tolerance. Beam segments are cropped.",fontsize=10)
    return fig


if __name__=="__main__":
    parser=argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--output",type=Path,default=OUT/"figure_2_4.png",
                        help="Single stacked-atlas output: PNG, PDF or SVG.")
    parser.add_argument("--panels",action="store_true",
                        help="Also save separate a-d and e-h panel PNGs beside the output.")
    args=parser.parse_args()
    with (OUT/"figure_2_4.csv").open(newline="",encoding="utf-8") as f:DATA=list(csv.DictReader(f))
    for metric in ["move_classification_mismatch_count","mpc_classification_mismatch_count"]:
        assert float(next(r["value"] for r in DATA if r["metric"]==metric))==0
    panels=[upper(DATA),lower(DATA)]
    rendered=[]
    for panel in panels:
        panel.set_dpi(220);panel.canvas.draw()
        rendered.append(np.asarray(panel.canvas.buffer_rgba()).copy())
    # Keep both source figures' geometry and scale intact: stack their native
    # raster canvases without stretching. PDF/SVG embed this same HD bitmap.
    width=max(image.shape[1] for image in rendered)
    height=sum(image.shape[0] for image in rendered)
    atlas=np.full((height,width,4),255,dtype=np.uint8)
    top=0
    for raster in rendered:
        atlas[top:top+raster.shape[0],:raster.shape[1]]=raster
        top+=raster.shape[0]
    composite=plt.figure(figsize=(width/220,height/220),dpi=220)
    ax=composite.add_axes([0,0,1,1]);ax.imshow(atlas);ax.axis("off")
    args.output.parent.mkdir(parents=True,exist_ok=True)
    composite.savefig(args.output,dpi=220,pad_inches=0)
    if args.panels:
        for suffix,panel in zip(["a","b"],panels):
            panel.savefig(args.output.with_name(args.output.stem+suffix+".png"),dpi=220)
    for panel in panels+[composite]:plt.close(panel)
    print(f"Rendered {args.output.name} from figure_2_4.csv only.")

File: data/figures/chapter02/figure_2_4/plot_2_4.m
function outputPath = plot_2_4(outputPath,savePanels)
%PLOT_2_4 Reproduce the eight-panel atlas using figure_2_4.csv only.
% Optional outputPath selects PNG/PDF/SVG. Optional savePanels=true also
% writes separate a-d/e-h PNGs beside that output. Default: one atlas PNG.
% Thick orange = lattice beam centre-lines; thin grey = skin-element edges;
% blue = receiving facets. Dashed proposals are not retained mesh members.
out=fileparts(mfilename('fullpath'));
if nargin<1 || isempty(outputPath),outputPath=fullfile(out,'figure_2_4.png');end
if nargin<2,savePanels=false;end
outputPath=char(outputPath);
[outputFolder,outputStem,extension]=fileparts(outputPath);
if ~any(strcmpi(extension,{'.png','.pdf','.svg'}))
    error('plot_2_4:OutputFormat','Output must use PNG, PDF or SVG.');
end
if ~isempty(outputFolder) && ~isfolder(outputFolder),mkdir(outputFolder);end
data=readtable(fullfile(out,'figure_2_4.csv'),'TextType','string');
panelA=upper_panels(data);
panelB=lower_panels(data);
% Render native figures directly to arrays: no panel image is an input and
% the atlas neither stretches coordinates nor changes any panel's scale.
imageA=print(panelA,'-RGBImage','-r220');
imageB=print(panelB,'-RGBImage','-r220');
close(panelA);close(panelB);
width=max(size(imageA,2),size(imageB,2));
height=size(imageA,1)+size(imageB,1);
atlas=uint8(255*ones(height,width,3));
atlas(1:size(imageA,1),1:size(imageA,2),:)=imageA;
atlas(size(imageA,1)+(1:size(imageB,1)),1:size(imageB,2),:)=imageB;
if strcmpi(extension,'.png')
    imwrite(atlas,outputPath);
else
    fig=figure('Visible','off','Color','w','Units','pixels', ...
        'Position',[80 80 width/2 height/2]);
    ax=axes(fig,'Position',[0 0 1 1]);image(ax,atlas);axis(ax,'image');axis(ax,'off');
    set(fig,'PaperUnits','inches','PaperPosition',[0 0 width/220 height/220], ...
        'PaperSize',[width/220 height/220]);
    if strcmpi(extension,'.pdf'),device='-dpdf';else,device='-dsvg';end
    print(fig,outputPath,device,'-r220');close(fig);
end
if savePanels
    imwrite(imageA,fullfile(outputFolder,[outputStem 'a.png']));
    imwrite(imageB,fullfile(outputFolder,[outputStem 'b.png']));
end
fprintf('Rendered eight panels to %s from one CSV.\n',outputPath);
end

function fig = upper_panels(data)
% Static scientific figure contract: two actual classification examples.
% Top: original/final single-B repair; bottom: rejected proposal/final mesh.
% Equal metric side-view scales, real neighboring skin facets; no magnification.
% Orange beam centre-lines, grey skin edges, blue actual receiving facets.
% A rejected proposal uses dashed orange lines and a cross, not a solid beam.
% Node and shell-vertex rows come from the single accompanying CSV file.
[nodes,shells]=upper_tables(data);
orange=[.78 .38 .12];blue=[.19 .39 .62];grey=[.53 .56 .59];ink=[.17 .19 .21];
fig=figure('Visible','off','Color','w','Position',[80 80 1400 680]);
set(fig,'DefaultAxesFontName','Arial','DefaultTextFontName','Arial');
names=["single_B_A_ordinary_C_beyond","B_rejected_length"];
titles={'(a) Single-node repair: original','(b) Single-node repair: final', ...
    '(c) Length guard: proposed correction','(d) Length guard: retained mesh'};
for row=1:2
    n=nodes(nodes.example==names(row),:);s=shells(shells.example==names(row),:);
    n=n([find(n.role=="A"),find(n.role=="B"),find(n.role=="C")],:);
    old=[n.original_x_m,n.original_y_m,n.original_z_m];
    final=[n.final_x_m,n.final_y_m,n.final_z_m];
    proposal=[n.all_proposals_x_m,n.all_proposals_y_m,n.all_proposals_z_m];
    p=[n.projection_x_m(2),n.projection_y_m(2),n.projection_z_m(2)];
    st=s(s.shell_element_id==n.selected_shell_element_id(2),:);
    v=[st.x_m,st.y_m,st.z_m];normal=cross(v(2,:)-v(1,:),v(3,:)-v(1,:));normal=normal/norm(normal);
    if dot(p-old(2,:),normal)<0,normal=-normal;end
    x=old(3,:)-old(1,:);x=x-dot(x,normal)*normal;x=x/norm(x);
    basis=[x(:),normal(:)];
    originalQ=(old-p)*basis*1000;finalQ=(final-p)*basis*1000;proposedQ=(proposal-p)*basis*1000;
    % Keep the one-ring facets, but bound the view around the shown beam chain.
    xmin=min(originalQ(:,1))-.75;xmax=max(originalQ(:,1))+.75;
    ymin=min(originalQ(:,2))-.7;ymax=.85;
    for col=1:2
        panel=(row-1)*2+col;
        ax=axes(fig,'Position',[.055+.475*(col-1),.56-.45*(row-1),.40,.30]);hold(ax,'on');
        accepted=n.category=="within_mpc_tolerance" | n.category=="snapped_to_shell_projection";
        receiving=unique(n.selected_shell_element_id(accepted));
        for eid=unique(s.shell_element_id).'
            tri=s(s.shell_element_id==eid,:);v=([tri.x_m,tri.y_m,tri.z_m]-p)*basis*1000;
            color=grey;width=.55;
            if ismember(eid,receiving),color=blue;width=1.4;end
            patch(ax,'Vertices',v,'Faces',[1 2 3],'FaceColor','none','EdgeColor',color,'LineWidth',width);
        end
        if row==1 && col==1,q=originalQ;style='-';labelB='B';
        elseif row==2 && col==1,q=proposedQ;style='--';labelB='B*';
        else,q=finalQ;style='-';labelB='B';if row==1,labelB='B''';end;end
        plot(ax,q(:,1),q(:,2),style,'Color',orange,'LineWidth',3);
        plot(ax,q([1 3],1),q([1 3],2),'o','MarkerFaceColor',ink,'MarkerEdgeColor','w','MarkerSize',7);
        marker='o';fill='w';if row==1 && col==2,fill=orange;end
        if row==2 && col==1,marker='x';end
        plot(ax,q(2,1),q(2,2),marker,'Color',orange,'MarkerEdgeColor',orange,'MarkerFaceColor',fill,'MarkerSize',9,'LineWidth',1.6);
        text(ax,q(1,1)-.3,q(1,2)-.22,'A','FontSize',12,'FontWeight','bold','Color',ink);
        text(ax,q(3,1)+.10,q(3,2)-.10,'C','FontSize',12,'FontWeight','bold','Color',ink);
        text(ax,q(2,1)-.1,q(2,2)-.35,labelB,'FontSize',12,'FontWeight','bold','Color',orange);
        axis(ax,'equal');xlim(ax,[xmin xmax]);ylim(ax,[ymin ymax]);set(ax,'Visible','off');
        title(ax,titles{panel},'Visible','on','FontSize',13,'FontWeight','normal');
        plot(ax,[xmin+.15,xmin+1.15],[ymin+.2,ymin+.2],'-','Color',ink,'LineWidth',1);
        text(ax,xmin+.65,ymin-.03,'1 mm','FontSize',9,'HorizontalAlignment','center');
    end
end
annotation(fig,'textbox',[.055 .462 .90 .042],'String', ...
    'B moves 1.701 mm; A is coupled without movement; C is outside repair reach (3.402 > 3.064 mm).', ...
    'LineStyle','none','FontSize',11,'Interpreter','none');
annotation(fig,'textbox',[.055 .02 .90 .042],'String', ...
    'B* is a rejected proposal: AB would lengthen by 81.9%, exceeding the 75% guard. B stays unchanged.', ...
    'LineStyle','none','FontSize',11,'Interpreter','none');
set(fig,'PaperPositionMode','auto');


end

function fig = lower_panels(data)
% Chart contract: show two distinct, observed near-skin coupling outcomes.
% Upper row: both adjacent lattice nodes relocate to their own skin facets.
% Lower row: an 8.189 micrometre gap remains when ordinary MPC coupling is
% permitted. The lower row is a uniformly magnified, cropped local side view.
% Reproduction requires only figure_2_4.csv, not a solver or archive.
% Explicit data-class encoding: thin grey skin edges, blue receiving facets,
% thick orange beam centre-lines. Open/filled markers denote original/final.
d=lower_table(data);
blue=[.19 .39 .62];orange=[.78 .38 .12];ink=[.17 .19 .21];grey=[.56 .60 .63];
fig=figure('Visible','off','Color','w','Position',[80 80 1400 730]);
set(fig,'DefaultAxesFontName','Arial','DefaultTextFontName','Arial');
u=d(d.example=="two_node",:);
target=u(u.entity=="receiving_vertex" & u.element_id==448671,:);
p=xyz(u(u.entity=="projection" & u.grid_id==491572,:));
a=xyz(u(u.entity=="beam_endpoint" & u.grid_id==491584 & u.state=="original",:));
c=xyz(u(u.entity=="beam_endpoint" & u.grid_id==491564 & u.state=="original",:));
v=xyz(target);n=cross(v(2,:)-v(1,:),v(3,:)-v(1,:));n=n/norm(n);
b=unique(xyz(u(u.entity=="beam_endpoint" & u.grid_id==491572 & u.state=="original",:)),'rows');
if dot(p-b,n)<0,n=-n;end
t=c-a;t=t-dot(t,n)*n;t=t/norm(t);s=cross(n,t);basis=[t(:),s(:),n(:)];
assert(norm(basis.'*basis-eye(3),'fro')<1e-12);
tr=@(rows) (xyz(rows)-p)*basis*1000;
all=tr(u);xlimTop=[min(all(:,1))-.3,max(all(:,1))+.3];
ylimTop=[min(all(:,2))-.2,max(all(:,2))+.2];zlimTop=[-3.15,.5];
% Upper row: oblique rendering of true mesh coordinates.
for panel=1:2
    ax=axes(fig,'Position',[.045+(panel-1)*.5 .555 .41 .37]);hold(ax,'on');
    state="original";if panel==2,state="final";end
    for eid=unique(u.element_id(u.entity=="context_vertex")).'
        q=tr(u(u.entity=="context_vertex" & u.element_id==eid,:));
        patch(ax,'Vertices',q,'Faces',[1 2 3],'FaceColor',[.86 .88 .90], ...
            'FaceAlpha',.10,'EdgeColor',grey,'LineWidth',.7);
    end
    for eid=[448671 449087]
        q=tr(u(u.entity=="receiving_vertex" & u.element_id==eid,:));
        patch(ax,'Vertices',q,'Faces',[1 2 3],'FaceColor',blue,'FaceAlpha',.22, ...
            'EdgeColor',blue,'LineWidth',1.2);
        centre=mean(q,1);name='T_B';dz=.30;if eid==449087,name='T_C';dz=-.60;end
        text(ax,centre(1),centre(2),centre(3)+dz,name, ...
            'Color',blue,'FontSize',12,'HorizontalAlignment','center','Tag','facet_label');
    end
    for eid=[829120 829109]
        q=tr(u(u.entity=="beam_endpoint" & u.element_id==eid & u.state==state,:));
        plot3(ax,q(:,1),q(:,2),q(:,3),'-','Color',orange,'LineWidth',3.2);
    end
    for gid=[491584 491572 491564]
        q=unique(tr(u(u.entity=="beam_endpoint" & u.grid_id==gid & u.state==state,:)),'rows');
        fill=ink;if gid~=491584,fill='w';if panel==2,fill=orange;end,end
        marker='o';if gid==491564,marker='d';end
        plot3(ax,q(1),q(2),q(3),marker,'MarkerFaceColor',fill,'MarkerEdgeColor',orange,'MarkerSize',8,'LineWidth',1.5);
        name='A';if gid==491572,name='B';elseif gid==491564,name='C';end
        if panel==2 && gid~=491584,name=[name ''''];end
        dz=.20;if gid==491572 && panel==2,dz=-.40;end
        if gid==491564 && panel==1,dz=-.2;end
        text(ax,q(1)+.08,q(2),q(3)+dz,name,'Color',ink,'FontSize',14,'FontWeight','bold');
    end
    if panel==1
        q=tr(u(u.entity=="projection",:));
        plot3(ax,q(:,1),q(:,2),q(:,3),'x','Color',blue,'MarkerSize',7,'LineWidth',1.2);
    end
    axis(ax,'equal');xlim(ax,xlimTop);ylim(ax,ylimTop);zlim(ax,zlimTop);
    view(ax,[10 25]);camproj(ax,'orthographic');camzoom(ax,1.25);set(ax,'Visible','off','Clipping','off');
    sx=-2.4;sy=0;sz=-2.7;
    plot3(ax,[sx sx+1],[sy sy],[sz sz],'-','Color',ink,'LineWidth',1.2);
    text(ax,sx+.5,sy,sz-.27,'1 mm','Color',ink,'FontSize',10,'HorizontalAlignment','center');
    if panel==1,heading='(e) Two adjacent nodes: before repair';else,heading='(f) Both nodes moved to their receiving facets';end
    note(fig,[.095+(panel-1)*.5 .948 .40 .04],heading,14,false,ink);
end
note(fig,[.055 .485 .89 .043], ...
    'B correction: 1.298 mm; C correction: 2.597 mm. Two receiving triangles are highlighted; their shared edge remains.',12,true,ink);
% Lower row: exact orthographic side coordinates in micrometres. Uniform
% zoom, equal length scales and beam clipping make the tiny offset visible.
u=d(d.example=="finite_offset",:);
p=xyz(u(u.entity=="projection",:));
b=unique(xyz(u(u.entity=="beam_endpoint" & u.grid_id==472299 & u.state=="original",:)),'rows');
v=xyz(u(u.entity=="receiving_vertex",:));
n=cross(v(2,:)-v(1,:),v(3,:)-v(1,:));n=n/norm(n);if dot(p-b,n)<0,n=-n;end
endrows=u(u.entity=="beam_endpoint" & u.grid_id~=472299 & u.state=="original",:);
w=xyz(endrows);t=w(2,:)-w(1,:);t=t-dot(t,n)*n;t=t/norm(t);
tr=@(rows) (xyz(rows)-p)*[t(:),n(:)]*1e6;
assert(abs(norm((b-p)*1e6)-8.1891470257)<1e-6);
for panel=1:2
    ax=axes(fig,'Position',[.070+(panel-1)*.5 .105 .365 .295]);hold(ax,'on');
    state="original";if panel==2,state="final";end
    % All receiving vertices project to n=0 in this true side view. Its
    % boundary appears as a line, not as a structural beam.
    q=tr(u(u.entity=="receiving_vertex",:));
    plot(ax,q([1:3 1],1),q([1:3 1],2),'-','Color',blue,'LineWidth',1.5);
    for eid=unique(u.element_id(u.entity=="beam_endpoint")).'
        q=tr(u(u.entity=="beam_endpoint" & u.element_id==eid & u.state==state,:));
        plot(ax,q(:,1),q(:,2),'-','Color',orange,'LineWidth',3.2);
    end
    q=unique(tr(u(u.entity=="beam_endpoint" & u.grid_id==472299 & u.state==state,:)),'rows');
    plot(ax,q(1),q(2),'o','MarkerFaceColor','w','MarkerEdgeColor',orange,'MarkerSize',8,'LineWidth',1.5);
    plot(ax,0,0,'x','Color',blue,'MarkerSize',8,'LineWidth',1.3);
    text(ax,-3.2,-9,'Q','Color',ink,'FontSize',14,'FontWeight','bold');
    text(ax,1.0,1.6,'Projection p','Color',blue,'FontSize',11);
    text(ax,-21,3.7,'Receiving skin element, edge-on','Color',blue,'FontSize',11);
    text(ax,4.0,-5.8,'Gap: 8.189 μm','Color',ink,'FontSize',11);
    % Dimension line intentionally omitted: the separated markers and label
    % show the gap without drawing anything that could be mistaken for a beam.
    axis(ax,'equal');xlim(ax,[-23 23]);ylim(ax,[-15 7]);
    plot(ax,[-21 -16],[-13 -13],'-','Color',ink,'LineWidth',1.3);
    text(ax,-18.5,-14.6,'5 μm','HorizontalAlignment','center','Color',ink,'FontSize',10);
    set(ax,'Visible','off','Clipping','on');
    if panel==1,heading='(g) Near-skin node before coupling';else,heading='(h) Ordinary MPC: coordinates unchanged';end
    title(ax,heading,'Visible','on','FontSize',14,'FontWeight','normal','Color',ink);
end
note(fig,[.055 .020 .89 .047], ...
    'Uniform side-view close-up: the 8.189 μm gap is below the 10 μm coupling tolerance. Beam segments are cropped.',11,false,ink);
set(findall(fig,'-property','Interpreter'),'Interpreter','none');
set(findall(fig,'Tag','facet_label'),'Interpreter','tex');
set(fig,'PaperPositionMode','auto');

end

function q=xyz(rows)
q=[rows.x_m rows.y_m rows.z_m];
end

function note(fig,pos,message,size,bold,color)
weight='normal';if bold,weight='bold';end
annotation(fig,'textbox',pos,'String',message,'FontName','Arial','FontSize',size, ...
    'FontWeight',weight,'Color',color,'EdgeColor','none','Interpreter','none');
end


function [nodes,shells]=upper_tables(data)
names=["single_B_A_ordinary_C_beyond","B_rejected_length"];
nodes=data(data.entity=="node" & data.state=="original" & ismember(data.example,names),:);
nodes.category=nodes.classification;
nodes.selected_shell_element_id=nodes.element_id;
for state=["original","final","projection","all_proposals"]
    coords=zeros(height(nodes),3);
    for i=1:height(nodes)
        q=data(data.entity=="node" & data.example==nodes.example(i) & ...
            data.grid_id==nodes.grid_id(i) & data.state==state,:);
        assert(height(q)==1);
        coords(i,:)=[q.x_m q.y_m q.z_m];
    end
    for k=1:3
        letters="xyz";letter=extractBetween(letters,k,k);
        nodes.(char(state+"_"+letter+"_m"))=coords(:,k);
    end
end
shells=data(data.entity=="skin_vertex" & ismember(data.example,names),:);
shells.shell_element_id=shells.element_id;
end

function d=lower_table(data)
keep=ismember(data.example,["two_node","finite_offset"]);
keep=keep & (data.entity=="beam_endpoint" | data.entity=="skin_vertex" | ...
    (data.entity=="node" & data.state=="projection"));
d=data(keep,:);
receiving=d.entity=="skin_vertex" & ~ismissing(d.receiving_for) & strlength(d.receiving_for)>0;
context=d.entity=="skin_vertex" & ~receiving;
d.entity(receiving)="receiving_vertex";
d.entity(context)="context_vertex";
d.entity(d.entity=="node")="projection";
end
